specs/agents/log-sending.md

```diff
@@ -0,0 +1,59 @@
+# Log sending
+
+### `log_sending` configuration
+
+**Warning**: experimental feature, may be subject to change until GA. Also, only a small subset of agents will provide it before GA.
+
+Controls the ability to send logs directly from the agent to APM server.
+
+|                |                 |
+|----------------|-----------------|
+| Valid options  | `true`, `false` |
+| Default        | `false`         |
+| Dynamic        | `true`          |
+| Central config | `true`          |
+
+The APM server only supports log events as of version 8.6+, thus trying to use this with an older version should
+issue a warning/error in the agent logs.
+
+### Log event format
+
+On the agent side, there are two ways to get an ECS-formatted log message from a log event:
+- The application already uses [ecs-logging](https://github.com/elastic/ecs-logging)
+- The agent embeds a copy of [ecs-logging](https://github.com/elastic/ecs-logging), which might also be used for [log reformatting](./log-reformatting.md).
+
+In both cases, the output of [ecs-logging](https://github.com/elastic/ecs-logging) can be reused as follows:
+
+```
+{"log":<ecs-formatted-log-event>}\n`
+```
+
+The ECS logging event `<ecs-formatted-log-event>` must not include an `EOL` character in order to preserve the ND-JSON
+format where each event is written to a single line.
+
+### Log event fields
+
+The ECS logging fields are the same as the ones defined in log reformatting:
+- [required fields](./log-reformatting.md#required-fields)
+- [recommended fields](./log-reformatting.md#recommended-fields)
+
+However, the values of `service.name` and `service.version` can be omitted as they are redundant to the values that are
+already sent in the [ND-JSON metadata](metadata.md). In the case where the formatted ECS log event already contains
+them it might be more efficient to send the event as-is than rewriting the event.
+
+### Agent log
+
+When `log_sending` option is enabled, agents may also send their own logs to APM server.
+
+Agents usually have internal debug/trace logging statements that allow to diagnose communication issues and serialized data
+sent to APM server. Special care must be taken to ensure that sending APM agent logs do not trigger an exponential loop
+of log events or excessively large log event.
+For APM agent logs, ignoring those log statements is an acceptable compromise, if there is any communication or 
+serialization issue with APM server it will already be logged for application traces, logs and metrics sent by the agent.
```

Comment:
Usually, when sending data agents often provide debug/trace log statements that help to diagnose what is sent on the wire for apm-server, so if there is any encoding or communication issue with apm-server it can be investigated by reading the logs.

When the agent send its own logs to apm-server and there is a bug in the agent, for example a serialization error, but it could be for any communication issue, then trying to send the debug/trace statement with the same code path that triggered the initial error is quite likely to go bad in multiple ways:
- with a verbose log level, if the agent logs the raw response from server, then the log payload will snowball into a very large JSON (it was the case with the Java agent) until the server limit (about 1Mb) is reached.
- with some agents the events are queued for sending, which means that trying to send logs should not create more log events in the queue.

So, what I meant with this sentence, is that :
- communication or serialization issues with apm-server will be logged in agent logs (in the usual file/stdout).
- when sending or processing agent log events, we should not "log about the logs" to prevent the behavior described above.
- if there is a general communication error with apm-server when sending application data (logs, metrics, traces, ...) then it will trigger enough debug/trace statements in the agent log (file/stdout) to investigate the issue.

Reply:
Thanks for the explanation. Perhaps this?

```suggestion
For APM agent logs, ignoring those log statements is an acceptable compromise -- diagnosis of agent errors in serializing or communicating with APM server may rely on local logging.
```